### PR TITLE
Fixed bug where headers were not 64-bit hex strings

### PR DESCRIPTION
**Pull request by kaisen**
Previously, the generated HTTP headers weren't always consistent with the trace/span id format. Sometimes plain int strings were used, and in other cases, hex strings that started with '0x' were used.

This PR ensures that HTTP headers are always 16-character 64-bit hex strings and cleans up the internal usage of trace/span ids.

[The zipkin spec on trace/span ids and HTTP headers.](https://twitter.github.io/zipkin/Instrumenting.html#communicating-trace-information)

### CODE REVIEW

File: pyramid_zipkin/request_helper.py [hunk 1]
```diff
@@ -53,14 +36,14 @@ def get_trace_id(request):
     completely random trace id.
 
     :param: current active pyramid request
-    :returns: string representaiton of zipkin trace id
+    :returns: a 64-bit hex string
     """
     if 'X-B3-TraceId' in request.headers:
         return request.headers['X-B3-TraceId']
     elif 'zipkin.trace_id_generator' in request.registry.settings:
         return request.registry.settings['zipkin.trace_id_generator'](request)
```
Comment (prat0318) [inline pyramid_zipkin/request_helper.py]: Can we put some validation on the result to ensure it is indeed 64bit hex string ^

File: pyramid_zipkin/request_helper.py [hunk 2]
```diff
@@ -147,8 +130,8 @@ def create_zipkin_attr(request):
 
     trace_id = request.zipkin_trace_id
     is_sampled = is_tracing(request)
-    span_id = request.headers.get('X-B3-SpanId', generate_span_id())
-    parent_span_id = request.headers.get('X-B3-ParentSpanId', '0')
+    span_id = request.headers.get('X-B3-SpanId', generate_random_64bit_string())
+    parent_span_id = request.headers.get('X-B3-ParentSpanId', '0000000000000000')
```
Comment (prat0318) [inline pyramid_zipkin/request_helper.py]: nit: '0' * 16

File: pyramid_zipkin/thrift_helper.py [hunk 3]
```diff
@@ -126,15 +126,15 @@ def create_span(zipkin_attrs, span_name, annotations, binary_annotations,
     :returns: zipkin span object
     """
 
-    span_id = generate_span_id() if is_client else zipkin_attrs.span_id
+    span_id = generate_random_64bit_string() if is_client \
```
Comment (prat0318) [inline pyramid_zipkin/thrift_helper.py]: nit: use () instead.
```
span_id =(gene..()
   if is_client else .. )
```

**Comment by prat0318** [comment]
lgtm other than minor nitpicks. Have we checked it with `yelp_lib` that the current `trace_id_generator` still works?

**Comment by kaisen** [comment]
`yelp_pyramid` will need to have `trace_id_generator` updated so that it returns the expected hex string.

**Comment by prat0318** [comment]
umm in that case, it will break services which have pyramid_zipkin unpinned. Is there a back-compat way?

**Comment by kaisen** [comment]
I added backwards compatibility so that services running the new version will still be able to process headers sent from older versions of pyramid_zipkin. This doesn't work in the reverse direction though, meaning services that send the new headers to a service with an older version of pyramid_zipkin will result in errors. We will have to roll this out starting with "leaf node" services.

File: docs/source/configuring_zipkin.rst [hunk 4]
```diff
@@ -104,8 +104,9 @@ fine tune as per your use case.
 
 5. zipkin.trace_id_generator
 ----------------------------
-    A method definition to generate a `trace_id` for the request. By default,
-    it creates a random trace id otherwise.
+    A method definition to generate a `trace_id` for the request. The trace_id is
+    expected to be a 64-bit hex string By default, it creates a random trace id
```
Comment (mjbryant) [inline docs/source/configuring_zipkin.rst]: Need a period between "hex string" and "By". I'd also like an example string, like "dfd0bda0f76a4c23", just to make sure people know it shouldn't have "0x" in front of it or anything.

File: pyramid_zipkin/request_helper.py [hunk 5]
```diff
@@ -28,23 +28,7 @@ class ZipkinAttrs(namedtuple(
 
 def generate_random_64bit_string():
     """Returns a 64 bit ascii encoded string"""
```
Comment (mjbryant) [inline pyramid_zipkin/request_helper.py]: Nit, but technically it's returning a UTF-8 encoded string. You can change the docstring, or you could use the ASCII codec to decode.

File: pyramid_zipkin/request_helper.py [hunk 6]
```diff
@@ -53,14 +37,36 @@ def get_trace_id(request):
     completely random trace id.
 
     :param: current active pyramid request
-    :returns: string representaiton of zipkin trace id
+    :returns: a 64-bit hex string
     """
     if 'X-B3-TraceId' in request.headers:
-        return request.headers['X-B3-TraceId']
+        id = request.headers['X-B3-TraceId']
```
Comment (mjbryant) [inline pyramid_zipkin/request_helper.py]: I'd avoid overloading the `id` variable. If someone, for whatever arcane reason, tries to access an object's location in memory later in this function it'd fail :P.

File: pyramid_zipkin/request_helper.py [hunk 7]
```diff
@@ -53,14 +37,36 @@ def get_trace_id(request):
...
+        id = generate_random_64bit_string()
+
+    # Backwards compatibility for <=v0.8.1
+    # If the trace id is a hex value that starts with '0x' or '-0x',
+    # convert to the unsigned form before proceeding.
+    if id.startswith('0x') or id.startswith('-0x'):
+        id = _signed_hex_to_unsigned_hex(id)
+    id = id.zfill(16)
+
+    assert _is_hex_string(id) and len(id) == 16
```
Comment (mjbryant) [inline pyramid_zipkin/request_helper.py]: I don't think you need to check that the len is 16 after you've `zfill`ed.

File: pyramid_zipkin/request_helper.py [hunk 8]
```diff
@@ -53,14 +37,36 @@ def get_trace_id(request):
...
+def _is_hex_string(s):
+    try:
+        int(s, 16)
+        return True
+    except ValueError:
+        return False
+
+
+def _signed_hex_to_unsigned_hex(s):
+    return '{0:x}'.format(struct.unpack('Q', struct.pack('q', int(s, 16)))[0])
```
Comment (mjbryant) [inline pyramid_zipkin/request_helper.py]: Would be great to give an example of the type of conversion this function is meant to apply.

File: pyramid_zipkin/thrift_helper.py [hunk 9]
```diff
@@ -153,14 +153,15 @@ def thrift_obj_in_bytes(thrift_obj):  # pragma: no cover
...
 
-    :param id_value: the string representation of a zipkin ID
-    :returns: int representation of zipkin ID
-    """
-    if len(id_value) == 0:
-        return 0
+    This is due to the fact that Apache Thrift only has signed values.
 
-    return int(id_value, 16)
+    :param hex_string: the string representation of a zipkin ID
```
Comment (mjbryant) [inline pyramid_zipkin/thrift_helper.py]: Example?

File: pyramid_zipkin/thrift_helper.py [hunk 10]
```diff
@@ -153,14 +153,15 @@ def thrift_obj_in_bytes(thrift_obj):  # pragma: no cover
...
+    :returns: signed int representation
```
Comment (mjbryant) [inline pyramid_zipkin/thrift_helper.py]: Example?

File: pyramid_zipkin/thrift_helper.py [hunk 11]
```diff
@@ -153,14 +153,15 @@ def thrift_obj_in_bytes(thrift_obj):  # pragma: no cover
...
+    """
+    # Backwards compatibility with ids generated by previous versions
+    if hex_string.startswith('-0x') or hex_string.startswith('0x'):
+        return int(hex_string, 16)
+    return struct.unpack('q', struct.pack('Q', int(hex_string, 16)))[0]
```
Comment (mjbryant) [inline pyramid_zipkin/thrift_helper.py]: Why can't you do `int(hex_string, 16)` in both cases? It looks to me like that works. I.e. `int('0xff', 16) == int('ff', 16)`, in both 2.7 and 3.4, AFAICT.

Comment (kaisen) [inline pyramid_zipkin/thrift_helper.py]: int() converts signed hex strings correctly but not unsigned hexes.

fail case:
int('b6dbb1c2b362bf51', 16) returns 13176320584593882961L but we want -5270423489115668655

File: pyramid_zipkin/request_helper.py [hunk 12]
```diff
@@ -53,14 +37,42 @@ def get_trace_id(request):
...
+    # If the trace id is a hex value that starts with '0x' or '-0x',
+    # convert to the unsigned form before proceeding.
+    if trace_id.startswith('0x') or trace_id.startswith('-0x'):
+        trace_id = _signed_hex_to_unsigned_hex(trace_id)
+    trace_id = trace_id.zfill(16)
+
+    return trace_id
+
+
+def _is_hex_string(s):
```
Comment (mjbryant) [inline pyramid_zipkin/request_helper.py]: You're not using this anywhere anymore.

**Comment by mjbryant** [comment]
Looks good to me! Delete the `_is_hex_string` function and it's ready to merge in.

**Pull request merged**